### ISSUE TITLE
QueryBuilder::buildSubqueryInCondition support multiple databases.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#15458]
Suppose I have 2 databases, A and B.
Database A has table `posts(id, author_id)`.
Database B has table `authors(id, name)`.
I have 2 models, `Post` and `Author`.

### What steps will reproduce the problem?
I'm writing a query like this:
````
Post::find()->where(['author_id' => Author::find()->where(['name' => 'Bob'])->select('id')]);
````

### What is the expected result?
Yii should detect that the databases are different and fallback to a multiple query approach. (The same it does for relationships.

### What do you get instead?
It fails because the subquery is not valid on the database used by the parent query.

### Additional info
To implement this some changes need to be made.
1. We add a `getDb()` and a `setDb()` function to `Query`. 
2. All database resolution is done via this database with a fallback to current functionality (ie in `ActiveQuery` it tries `model::getDb()` and then in `Query` it tries `app->db`.
3. This will also allow AR to support it for mssql and sqlite, currently with those drivers an exception is thrown since for multicolumn in clause subqueries.

I think functionally both methods will always yield the same results, only one is more performant than the other. In cases where the database is different we can assume that the developer is aware of the performance loss and has made the choice based on other parameters.

**Comment by samdark** [comment]
Sounds interesting but also sounds complicated. Not sure it could be done in 2.0 and a bit unsure if making query connection-aware is a good idea overall. @yiisoft/core-developers, @yiisoft/reviewers need your opinions.

**Comment by SamMousa** [comment]
It doesn't need to be connection aware with respect to implementation. It just would do strict object comparison. Blindly assuming they're the same (current implementation) seems less logical to me :)

**Comment by samdark** [comment]
Connection-aware means it will store a link to connection instance.

**Comment by SamMousa** [comment]
Hmm, it is aware off the global service locator and it's database service.

**Comment by samdark** [comment]
How would `getDb/setDb` look like?

**Comment by SamMousa** [comment]
For a normal query:
````
protected $_db;
public function getDb() 
{
    return $this->_db ?? \Yii::$app->db();
}
````
For AQ:
````
public function getDb() 
{
    $modelClass = $this->modelClass;
    $db = $modelClass::getDb();
    return $this->_db ?? $modelClass::getDb() ?? parent::getDb();
}
````

I don't have a use case for `setDb()` at this time; so it could be omitted. (Though I think at some point someone will come up with one.)

**Comment by samdark** [comment]
How would it receive connection instance? i.e. when `$this->_db` becomes not null?

**Comment by SamMousa** [comment]
Via a setter, that is called in the query methods that have a db arg, for now..

**Comment by samdark** [comment]
Won't be implemented in 2.0.
